frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -587,18 +604,87 @@ impl<T: Config> Pallet<T> {
 						Ok(now)
 					}
 				}
-			});
+			},
+		);
 
 		match mutate_stat {
 			// all good
 			Ok(Ok(_)) => Ok(()),
 			// failed to write.
-			Ok(Err(_)) => Err(MinerError::Lock("failed to write to offchain db.")),
+			Ok(Err(_)) => Err(MinerError::Lock("failed to write to offchain db [repeat].")),
 			// fork etc.
 			Err(why) => Err(MinerError::Lock(why)),
 		}
 	}
 
+	/// Ensure a new offchain worker is not being executed at this point in time, by checking
+	/// `OFFCHAIN_RUNNING` storage.
+	///
+	/// If `OK(_)` is returned, `true` is written to `OFFCHAIN_RUNNING`, which in turn prevents any
+	/// further call to this function to succeed.
+	///
+	/// It is up to the call site, and ABSOLUTELY CRUCIAL to make sure `release_offchain_lock` is
+	/// called appropriately to ensure starvation does not occur.
+	pub(crate) fn ensure_offchain_not_running() -> Result<(), MinerError> {
+		let guard = StorageValueRef::persistent(&OFFCHAIN_RUNNING);
+
+		let mutate_stat =
+			guard.mutate::<_, &'static str, _>(|maybe_running: Option<Option<bool>>| {
+				match maybe_running {
+					Some(Some(false)) => {
+						// no one else is running. Set it, and return Ok.
+						Ok(true)
+					}
+					Some(Some(true)) => {
+						// another thread is currently running.
+						Err("still running.")
+					}
+					_ => {
+						// value does not exist.
+						Ok(true)
+					}
+				}
+			});
+
+		match mutate_stat {
+			// all good
+			Ok(Ok(_)) => Ok(()),
+			// failed to write.
+			Ok(Err(_)) => Err(MinerError::Lock("failed to write to offchain db [guard].")),
+			// still running etc.
+			Err(why) => Err(MinerError::Lock(why)),
+		}
+	}
+
+	/// Perform all the checks that are needed before we spawn any significant work in the offchain
+	/// worker thread.
+	///
+	/// This tries first `ensure_offchain_not_running` and then `ensure_offchain_repeat_frequency`.
+	pub(crate) fn try_acquire_offchain_lock(now: T::BlockNumber) -> Result<(), MinerError> {
+		Self::ensure_offchain_not_running().and_then(|_| {
+			Self::ensure_offchain_repeat_frequency(now).map_err(|error| {
+				// don't alter the error, just release the lock acquired in
+				// `ensure_offchain_not_running` if we fail this second step.
+				Self::release_offchain_lock();
+				error
+			})
+		})
+	}
+
+	/// Perform the release operation of any locks that where acquired to execute the offchain
+	/// worker thread.
+	pub(crate) fn release_offchain_lock() {
+		log!(debug, "releasing offchain repeat lock.");
+		let guard = StorageValueRef::persistent(&OFFCHAIN_RUNNING);
+		let _mutate_stats = guard.mutate::<_, &'static str, _>(|_| Ok(false));
+		// defensive-only: ^^ it can only fail if another offchain thread mutates the
```

Comment:
@tomusdrw can you sign off on this? If I am correct, we might as well just use the racy `set`.

Reply:
Yes, that's about right. It can also fail if the stored value encoding does not match the expected one, but I think the error would be a bit different.

You can also get storage modifications from RPC or externally, but that's an edge case. I'd use `set` here (or consider using `StorageLock` structure to handle abrupt execution error).